<a href="https://colab.research.google.com/github/faisu6339-glitch/Deep-Learning/blob/main/integer_encoding_simplernn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np

docs = ['go india',
		'india india',
		'hip hip hurray',
		'jeetega bhai jeetega india jeetega',
		'bharat mata ki jai',
		'kohli kohli',
		'sachin sachin',
		'dhoni dhoni',
		'modi ji ki jai',
		'inquilab zindabad']

In [ ]:
from keras.preprocessing.text import Tokenizer
tokenizer = Tokenizer(oov_token='<nothing>')

In [ ]:
tokenizer.fit_on_texts(docs)

In [ ]:
tokenizer.word_index

{'<nothing>': 1,
 'india': 2,
 'jeetega': 3,
 'hip': 4,
 'ki': 5,
 'jai': 6,
 'kohli': 7,
 'sachin': 8,
 'dhoni': 9,
 'go': 10,
 'hurray': 11,
 'bhai': 12,
 'bharat': 13,
 'mata': 14,
 'modi': 15,
 'ji': 16,
 'inquilab': 17,
 'zindabad': 18}

In [ ]:
tokenizer.word_counts

OrderedDict([('go', 1),
             ('india', 4),
             ('hip', 2),
             ('hurray', 1),
             ('jeetega', 3),
             ('bhai', 1),
             ('bharat', 1),
             ('mata', 1),
             ('ki', 2),
             ('jai', 2),
             ('kohli', 2),
             ('sachin', 2),
             ('dhoni', 2),
             ('modi', 1),
             ('ji', 1),
             ('inquilab', 1),
             ('zindabad', 1)])

In [ ]:
tokenizer.document_count

10

In [ ]:
sequences = tokenizer.texts_to_sequences(docs)
sequences

[[10, 2],
 [2, 2],
 [4, 4, 11],
 [3, 12, 3, 2, 3],
 [13, 14, 5, 6],
 [7, 7],
 [8, 8],
 [9, 9],
 [15, 16, 5, 6],
 [17, 18]]

In [ ]:
from keras.utils import pad_sequences

In [ ]:
sequences = pad_sequences(sequences,padding='post')

In [ ]:
sequences

array([[10,  2,  0,  0,  0],
       [ 2,  2,  0,  0,  0],
       [ 4,  4, 11,  0,  0],
       [ 3, 12,  3,  2,  3],
       [13, 14,  5,  6,  0],
       [ 7,  7,  0,  0,  0],
       [ 8,  8,  0,  0,  0],
       [ 9,  9,  0,  0,  0],
       [15, 16,  5,  6,  0],
       [17, 18,  0,  0,  0]], dtype=int32)

In [ ]:
from keras.datasets import imdb
from keras import Sequential
from keras.layers import Dense,SimpleRNN,Embedding,Flatten

In [ ]:
(X_train,y_train),(X_test,y_test) = imdb.load_data()

17464789/17464789 [==============================] - 1s 0us/step


In [ ]:
X_train[0]

[1,
 14,
 22,
 16,
 43,
 530,
 973,
 1622,
 1385,
 65,
 458,
 4468,
 66,
 3941,
 4,
 173,
 36,
 256,
 5,
 25,
 100,
 43,
 838,
 112,
 50,
 670,
 22665,
 9,
 35,
 480,
 284,
 5,
 150,
 4,
 172,
 112,
 167,
 21631,
 336,
 385,
 39,
 4,
 172,
 4536,
 1111,
 17,
 546,
 38,
 13,
 447,
 4,
 192,
 50,
 16,
 6,
 147,
 2025,
 19,
 14,
 22,
 4,
 1920,
 4613,
 469,
 4,
 22,
 71,
 87,
 12,
 16,
 43,
 530,
 38,
 76,
 15,
 13,
 1247,
 4,
 22,
 17,
 515,
 17,
 12,
 16,
 626,
 18,
 19193,
 5,
 62,
 386,
 12,
 8,
 316,
 8,
 106,
 5,
 4,
 2223,
 5244,
 16,
 480,
 66,
 3785,
 33,
 4,
 130,
 12,
 16,
 38,
 619,
 5,
 25,
 124,
 51,
 36,
 135,
 48,
 25,
 1415,
 33,
 6,
 22,
 12,
 215,
 28,
 77,
 52,
 5,
 14,
 407,
 16,
 82,
 10311,
 8,
 4,
 107,
 117,
 5952,
 15,
 256,
 4,
 31050,
 7,
 3766,
 5,
 723,
 36,
 71,
 43,
 530,
 476,
 26,
 400,
 317,
 46,
 7,
 4,
 12118,
 1029,
 13,
 104,
 88,
 4,
 381,
 15,
 297,
 98,
 32,
 2071,
 56,
 26,
 141,
 6,
 194,
 7486,
 18,
 4,
 226,
 22,
 21,
 134,
 476,
 26,
 480,
 5

In [ ]:
len(X_train[2])

141

In [ ]:
X_train = pad_sequences(X_train,padding='post',maxlen=50)
X_test = pad_sequences(X_test,padding='post',maxlen=50)

In [ ]:
X_train[0]

array([2071,   56,   26,  141,    6,  194, 7486,   18,    4,  226,   22,
         21,  134,  476,   26,  480,    5,  144,   30, 5535,   18,   51,
         36,   28,  224,   92,   25,  104,    4,  226,   65,   16,   38,
       1334,   88,   12,   16,  283,    5,   16, 4472,  113,  103,   32,
         15,   16, 5345,   19,  178,   32], dtype=int32)

In [ ]:
model = Sequential()

model.add(SimpleRNN(32,input_shape=(50,1),return_sequences=False))
model.add(Dense(1,activation='sigmoid'))

model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 simple_rnn (SimpleRNN)      (None, 32)                1088      
                                                                 
 dense (Dense)               (None, 1)                 33        
                                                                 
Total params: 1,121
Trainable params: 1,121
Non-trainable params: 0
_________________________________________________________________


In [ ]:
model.compile(loss='binary_crossentropy',optimizer='adam',metrics=['accuracy'])

model.fit(X_train,y_train,epochs=5,validation_data=(X_test,y_test))

Epoch 1/5
782/782 [==============================] - 10s 12ms/step - loss: 0.6929 - accuracy: 0.5061 - val_loss: 0.6955 - val_accuracy: 0.5048
Epoch 2/5
782/782 [==============================] - 9s 12ms/step - loss: 0.6931 - accuracy: 0.5031 - val_loss: 0.6945 - val_accuracy: 0.5021
Epoch 3/5
782/782 [==============================] - 9s 12ms/step - loss: 0.6927 - accuracy: 0.5085 - val_loss: 0.6967 - val_accuracy: 0.5008
Epoch 4/5
782/782 [==============================] - 9s 12ms/step - loss: 0.6927 - accuracy: 0.5090 - val_loss: 0.6942 - val_accuracy: 0.5043
Epoch 5/5
782/782 [==============================] - 9s 12ms/step - loss: 0.6928 - accuracy: 0.5056 - val_loss: 0.6957 - val_accuracy: 0.5048


## Explanation: Integer Encoding and SimpleRNN

This notebook demonstrates a basic workflow for preparing text data and training a simple recurrent neural network (RNN) for a classification task, likely sentiment analysis, using the IMDB movie review dataset. Let's break down the key concepts:

### 1. Integer Encoding (Tokenization)

The first step in processing text data for neural networks is to convert words into numerical representations. This is often done through **integer encoding**, where each unique word in your vocabulary is assigned a unique integer ID. The `keras.preprocessing.text.Tokenizer` is a powerful tool for this.

*   **`tokenizer = Tokenizer(oov_token='<nothing>')`**:
    *   Initializes the `Tokenizer`. The `oov_token` (Out-Of-Vocabulary token) is crucial. If set, any word encountered during `texts_to_sequences` that was not seen during `fit_on_texts` will be replaced by this token's ID. In your case, it's set to `<nothing>`, and it gets assigned the first index (1) because it's effectively treated as the most frequent 'word' to reserve its spot.

In [5]:
from tensorflow.keras.preprocessing.text import Tokenizer
# Re-initializing for demonstration purposes
demo_tokenizer = Tokenizer(oov_token='<nothing>')
demo_tokenizer.fit_on_texts(docs)
print('Word Index:', demo_tokenizer.word_index)
print('Word Counts:', demo_tokenizer.word_counts)
print('Document Count:', demo_tokenizer.document_count)

NameError: name 'docs' is not defined

*   **`tokenizer.fit_on_texts(docs)`**:
    *   This method builds the vocabulary from your input `docs` (a list of sentences). It scans all the words, assigns unique integers to them based on their frequency (more frequent words get lower integer IDs, though `oov_token` usually takes 1), and updates internal dictionaries like `word_index` and `word_counts`.

*   **`tokenizer.word_index`**:
    *   A dictionary that maps each word to its assigned integer ID. You can see from the output that `india` got ID 2 (most frequent non-OOD), `jeetega` got 3, and so on.

*   **`tokenizer.word_counts`**:
    *   A dictionary that stores the frequency of each word in the corpus.

*   **`tokenizer.document_count`**:
    *   The total number of documents (sentences in your `docs` list) processed.

*   **`sequences = tokenizer.texts_to_sequences(docs)`**:
    *   This is where the actual integer encoding happens. Each sentence in `docs` is converted into a list of integers, where each integer corresponds to the word's ID from `word_index`.

In [2]:
sequences = tokenizer.texts_to_sequences(docs)
print('Encoded sequences:', sequences)

NameError: name 'tokenizer' is not defined

### 2. Padding Sequences

Neural networks, especially RNNs, often expect input sequences of uniform length. However, sentences (and thus their integer-encoded sequences) typically have varying lengths. This is where padding comes in.

*   **`from keras.utils import pad_sequences`**:
    *   Imports the utility function for padding.

*   **`sequences = pad_sequences(sequences, padding='post', maxlen=50)`**:
    *   `pad_sequences` takes your list of integer sequences and makes them all the same length.
    *   `padding='post'` means that zeros will be added to the *end* of shorter sequences to reach the desired length. If it were `'pre'`, zeros would be added to the beginning.
    *   `maxlen` (used for the IMDB dataset) specifies the target length for all sequences. If a sequence is longer than `maxlen`, it will be truncated. If it's shorter, it will be padded. In your IMDB example, `maxlen=50` ensures all movie reviews are represented by a sequence of 50 integers.

In [3]:
from keras.utils import pad_sequences

# Example with your 'docs' sequences
# Let's say we want all sequences to be length 5 for demonstration
padded_demo_sequences = pad_sequences(sequences, padding='post', maxlen=5)
print('Padded demo sequences:', padded_demo_sequences)

# This is what happened with your IMDB data (X_train and X_test)
# The original sequences (movie reviews) were padded/truncated to a length of 50.

NameError: name 'sequences' is not defined

### 3. IMDB Dataset

*   **`(X_train, y_train), (X_test, y_test) = imdb.load_data()`**:
    *   This loads the IMDB movie review sentiment classification dataset.
    *   `X_train` and `X_test` contain pre-processed integer sequences representing movie reviews. These integers are already mapped to words based on their frequency in the IMDB corpus.
    *   `y_train` and `y_test` contain the corresponding labels (0 for negative sentiment, 1 for positive sentiment).
    *   The `imdb.load_data()` function automatically handles a lot of the tokenization and integer mapping for you, but the padding step is still necessary to unify sequence lengths.

### 4. SimpleRNN Model

This section defines and summarizes a simple Recurrent Neural Network (RNN) model using Keras. RNNs are a type of neural network designed to process sequential data, making them suitable for natural language processing tasks.

*   **`from keras import Sequential`**: Imports the `Sequential` model API, which allows you to build models layer-by-layer.

*   **`from keras.layers import Dense, SimpleRNN, Embedding, Flatten`**: Imports the necessary layers.
    *   `Dense`: A standard fully connected neural network layer.
    *   `SimpleRNN`: The basic recurrent layer. It processes sequences one step at a time, maintaining an internal state that captures information from previous steps.
    *   `Embedding`: (Not used in *this specific* model's `add` calls, but commonly used for text. It would convert integer word IDs into dense vector representations. Since your `X_train` is already integer sequences, and `SimpleRNN` here is receiving `input_shape=(50,1)`, it treats each integer as a single feature, which is less common for text than using an `Embedding` layer first.)
    *   `Flatten`: (Not used in this model either, but would flatten the output of a previous layer into a 1D array.)

Let's look at your model definition:

*   **`model = Sequential()`**: Creates an empty sequential model.

*   **`model.add(SimpleRNN(32, input_shape=(50,1), return_sequences=False))`**:
    *   Adds a `SimpleRNN` layer.
    *   `32`: This is the number of recurrent units (neurons) in the layer, also known as the dimensionality of the output space (the hidden state).
    *   `input_shape=(50,1)`: This is crucial. It tells the RNN what to expect as input.
        *   `50`: The length of each sequence (the `maxlen` from `pad_sequences`).
        *   `1`: The number of features per time step. In this case, since you didn't use an `Embedding` layer, the model is treating each integer ID in the sequence as a single feature. This is a very basic way to feed integer-encoded data and might not capture semantic relationships between words as well as an `Embedding` layer would.
    *   `return_sequences=False`: This means the RNN layer will output only the last hidden state for each input sequence. If it were `True`, it would output the hidden state for every time step (all 50 steps), resulting in a 3D output.

*   **`model.add(Dense(1, activation='sigmoid'))`**:
    *   Adds a fully connected `Dense` layer.
    *   `1`: There is one neuron in this layer, suitable for binary classification (sentiment: positive/negative).
    *   `activation='sigmoid'`: The sigmoid activation function is used. It squashes the output to a value between 0 and 1, which can be interpreted as a probability for the positive class.

*   **`model.summary()`**:
    *   Prints a summary of the model's architecture, including the output shape of each layer and the number of parameters. You can see the `SimpleRNN` layer has 1088 parameters and the `Dense` layer has 33.

In [4]:
from keras import Sequential
from keras.layers import Dense,SimpleRNN,Embedding,Flatten

model = Sequential()

# Re-demonstrate the model creation
model.add(SimpleRNN(32,input_shape=(50,1),return_sequences=False))
model.add(Dense(1,activation='sigmoid'))

model.summary()

/usr/local/lib/python3.12/dist-packages/keras/src/layers/rnn/rnn.py:199: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ simple_rnn (SimpleRNN)          │ (None, 32)             │         1,088 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,121 (4.38 KB)

 Trainable params: 1,121 (4.38 KB)

 Non-trainable params: 0 (0.00 B)

### 5. Model Compilation and Training

*   **`model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])`**:
    *   **Compilation** is the step where you configure the learning process of your model.
    *   `loss='binary_crossentropy'`: This is the standard loss function for binary classification problems where the output is a probability between 0 and 1. It measures how well the model's predicted probabilities match the true labels.
    *   `optimizer='adam'`: Adam is a popular and effective optimization algorithm used to update the model's weights during training. It adapts the learning rate for each parameter, often leading to faster convergence.
    *   `metrics=['accuracy']`: This specifies the metric(s) to be evaluated by the model during training and testing. Accuracy measures the proportion of correctly classified samples.

*   **`model.fit(X_train, y_train, epochs=5, validation_data=(X_test, y_test))`**:
    *   **Training** is the core process where the model learns from the data.
    *   `X_train`, `y_train`: The training data and corresponding labels.
    *   `epochs=5`: The model will iterate over the entire training dataset 5 times. In each epoch, the model adjusts its weights to minimize the loss function.
    *   `validation_data=(X_test, y_test)`: During training, the model will also evaluate its performance (loss and metrics) on this separate validation dataset at the end of each epoch. This helps to monitor for overfitting – if the training loss continues to decrease but validation loss starts to increase, the model might be overfitting to the training data.

### Summary of Results

From your `model.fit` output:

*   The model achieves an accuracy of around 50-51% on both the training and validation sets. This indicates that the model is performing roughly at chance level (like flipping a coin for each prediction).
*   This low accuracy suggests that this very simple `SimpleRNN` with a `(50,1)` input shape (treating integer IDs as direct features without an embedding layer) is not sufficient for the complexity of the IMDB sentiment analysis task. Usually, an `Embedding` layer is essential to learn meaningful representations of words before feeding them into an RNN.

In [6]:
from tensorflow.keras.preprocessing.text import Tokenizer
tokenizer = Tokenizer(oov_token='<nothing>')